index.js

```diff
@@ -106,6 +118,11 @@
     return f(this.value);
   };
 
+  //  Just#concat :: m a -> m a
+  Just.prototype.concat = function(maybe) {
+    return maybe instanceof Just ? Just(this.value.concat(maybe.value)) : this;
+  };
+
```

Comment:
> No, but I was hoping that Maybe could be. My reason for adding Maybe to a system that otherwise doesn't have one is to prevent having to do all sorts of error-checking across the system, or, in failing to do that, allowing for the possibility of exceptions arising. So I want my Maybe to be as bullet-proof as possible.

That's an admirable goal. I think the only way to really make that happen is to stop writing vanilla js, or use a different abstraction. `Maybe` is too low an abstraction to protect you from all the possible ways things can go wrong. That doesn't mean `Maybe` is worthless, just that it's not up to the task for js. Different languages can use it more powerfully than js though, for the exact reasons you pointed out above. If you have a computer checking your program, you can ensure that you only call `concat` on the right thing. In js, you can't have that.


> But I finally went back and looked at the filter implementation involved, and perhaps I've been confused by the discussion. It looks as though filter has nothing to do with Semigroups or Monoids. There seems to be what might be seen to be a pun on empty here. There seems to be nothing in the definition of filter, explicit or implied, that requires empty be a lawful Monoid identity. filter only requires that such a method exists (and presumably requires that it returns a value of the same Maybe.) Moreover all my concerns above had to do with the definition of concat, and the fact that it could easily throw unexpected exceptions. But that's entirely irrelevant, since the definition of filter being discussed has nothing at all to do with the Maybe being a Semigroup or a Monoid.

This all sprung from my suggestion here: https://github.com/plaid/sanctuary/pull/10#issuecomment-74254613 When I said, "once and for all", I truly meant "once and for all". There shouldn't need to be an **implementation** of `filter` on each data type. If you have a data type that is a `Monad` and a `Monoid`, that interact properly, you should be able to use the a single filter implementation:

```javascript
var filter = function(p, m) {
  return m.chain(function(x) {
    return p(x) ? m.of(x) : m.empty();
  });
}
```

That method doesn't mention `Maybe a` or `[a]`, or any other data type. It talks only about `Monad` and `Monoid`. That implementation is the power of talking about these abstract concepts. We needn't think about, "Does it make sense to `filter` this data type?" or, "How do I implement a `filter` on this data type?" No, we only need to be concerned with some questions. does this data type implement `Monad`? Does this data type implement `Monoid`? Do they interact properly? If the answer to all of those questions is yes, then we get `filter` for free. You stick that implementation in some generalized library somewhere, and use it on your data type without even thinking about it.

But what does it mean to interact properly? There's some discussion here: https://github.com/purescript/purescript-control/issues/6#issuecomment-44097271, but the idea is that it should satisfy some laws:

* Left distributivity with `Functor`: 
```javascript
(x.concat(y)).map(f) == x.map(f).concat(y.map(f))
```
* Right annihilation with `Functor`
```javascript
M.empty().map(f) = M.empty()
```
* Right distributivity with `Apply`: 
```javascript
(x.concat(y)).ap(z) == x.ap(z).concat(y.ap(z))
```
* Right annihilation with `Apply`
```javascript
M.empty().ap(z) = M.empty()
```
* Right distributivity with `Chain`: 
```javascript
(x.concat(y)).chain(f) == x.chain(f).concat(y.chain(f))
```
* Left annihilation with `Chain`
```javascript
M.empty().chain(f) = M.empty()
```
Since `Maybe a` happens to satisfy these laws, it should get filter for free. It has nothing to do with HOW `Maybe a` implements these functions. Just that they're lawful.